idb/postgres/internal/util/util.go

```diff
@@ -0,0 +1,69 @@
+package util
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/lib/pq"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/algorand/indexer/idb"
+)
+
+// TxWithRetry is a helper function that retries the function `f` in case the database
+// transaction in it fails due to a serialization error. `f` is provided context `ctx`
+// and a transaction created using this context and `opts`. `f` takes ownership of the
+// transaction and must either call sql.Tx.Rollback() or sql.Tx.Commit(). In the second
+// case, `f` must return an error which contains the error returned by sql.Tx.Commit().
+// The easiest way is to just return the result of sql.Tx.Commit().
+func TxWithRetry(ctx context.Context, db *sql.DB, opts sql.TxOptions, f func(context.Context, *sql.Tx) error, log *log.Logger) error {
+	count := 0
+	for {
+		tx, err := db.BeginTx(ctx, &opts)
+		if err != nil {
+			return err
+		}
+
+		err = f(ctx, tx)
+
+		// If not serialization error.
+		var pqerr *pq.Error
+		if !errors.As(err, &pqerr) || (pqerr.Code != "40001") {
+			if (count > 0) && (log != nil) {
+				log.Printf("transaction was retried %d times", count)
+			}
+			return err
+		}
+
+		count++
```

Comment:
need to have a limit on retry count

Reply:
Maybe. But this is out of scope.

Reply:
I would like to avoid making assumptions about timing and this is out of scope.

Reply:
I agree with @brianolson that we definitely want to have this. however, this should not block merging this code in, since it's not any worse than prior to this change. @tolikzinovyev - could you please open a ticket for adding this technical debt ?

Reply:
https://github.com/algorand/indexer/issues/635